java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/UIDefaultsIconMetadata.java

```diff
@@ -0,0 +1,357 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.nbcode.integration;
+
+import java.awt.Component;
+import java.awt.EventQueue;
+import java.awt.Graphics;
+import java.awt.HeadlessException;
+import java.awt.Image;
+import java.awt.MediaTracker;
+import java.awt.Toolkit;
+import java.awt.Transparency;
+import java.awt.image.BufferedImage;
+import java.awt.image.ColorModel;
+import java.awt.image.ImageObserver;
+import java.awt.image.WritableRaster;
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.Enumeration;
+import java.util.Hashtable;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.swing.Icon;
+import javax.swing.ImageIcon;
+import javax.swing.UIDefaults;
+import javax.swing.UIDefaults.ActiveValue;
+import javax.swing.UIDefaults.LazyValue;
+import javax.swing.UIManager;
+import org.openide.modules.OnStart;
+import org.openide.util.ImageUtilities;
+
+/**
+ * Goes through UIDefaults and replaces Icons and Images with variants that
+ * return image's URL from the 'url' property.
+ * @author sdedic
+ */
+@OnStart
+public class UIDefaultsIconMetadata implements Runnable {
+    private static final Logger LOG = Logger.getLogger(UIDefaultsIconMetadata.class.getName());
+    
+    // for diagnostic purposes only
+    public static final String PROP_ORIGINAL_ICON = "originalIcon"; // NOI18N
+    // for diagnostic purposes only
+    public static final String PROP_ORIGINAL_IMAGE = "originalImage"; // NOI18N
+    
+    private static final String URN_DEFAULTS_PREFIX = "uidefaults:"; // NOI18N
+    
+    private static Object wrapImage(String key, Image image, Icon icon, Object... meta) {
+        Object o = image.getProperty(ImageUtilities.PROPERTY_ID, null); // NOI18N
+        if (o instanceof URL) {
+            return image;
+        }
+        
+        o = image.getProperty(ImageUtilities.PROPERTY_ID, null);
+        if (o instanceof String || o instanceof URI) {
+            return image;
+        } else {
+            
+            Hashtable props = new Hashtable();
+            for (int i = 0; i < meta.length; i+= 2) {
+                props.put(meta[i], meta[i+1]);
+            }
+            props.put(ImageUtilities.PROPERTY_ID, URN_DEFAULTS_PREFIX + key);
+            return createNew(image, icon, props);
+        }
+    }
+    
+    @Override
+    public void run() {
+        EventQueue.invokeLater(() -> {
+            replaceIconsAndImages(UIManager.getDefaults());
+        });
+    }
+    
+    static void replaceIconsAndImages(UIDefaults defs) {
```

Comment:
Could you explain why the UIDefaultsIconMetadata wrapping is necessary? As far as I know, icons that are loaded from files already have the "url" property set by ImageUtilities.

There's a lot of LAF customizations (and hacks) that interact with UIManager, and hooking up to it this way seems fragile and likely to cause problems in the future. Furthermore, the Icon wrapping logic in ImageUtilities is already hard to understand and debug, and I'd hate to wrap all the Icon instances in yet another layer...

Reply:
My goal was to provide "some identification" for an Image now exported to LSP, so the client can decide what to do with it. Ideally without changing **all** the places where an image is used or loaded. In addition images are loaded from UIDefaults by Swing as well (which I can not change). 

It may be the case that eventually I'll go through NB codebase and change the UIManager.getIcon() calls one by one (after I make up some alternative and for a SPI)... then this wrapper code code would eventually vanish without impact on its clients.

This place is at this moment the central and only one that knows, for a image, some ID - the hashmap key.

Reply:
I share Eirik's concerns, and also have concerns about changing the UIManager calls without some coordination. An API / SPI approach is certainly where I'd like to see this evolve, but that overlaps with a bunch of other conversations going on at the moment (inc. hidpi, svg, dark icons, branding). As you said before, there's a discussion to be had on dev@ about this. I still think this PR is potentially addressing the issue in the wrong place, but also that it's a case of needs must right now.

Reply:
@neilcsmith-net Oh, I always assumed that this PR was part of the effort to let icons be branded for LAF purposes. Is there any hope of satisfying both needs with one API change?

> It may be the case that eventually I'll go through NB codebase and change the UIManager.getIcon() calls one by one (after I make up some alternative and for a SPI)... then this wrapper code code would eventually vanish without impact on its clients.

Perhaps this might actually be preferable to hacking UIManager. How many of the icons would actually require this? It's only the select few that don't have an URL to a PNG or GIF for some reason, right?

Reply:
@eirikbakke note that this 'replacement' solves also icons queried by Swing codebase that might flow through the system: even if Swing calls `getIcon()`, it gets a wrapper ... and if that image/icon is passed around the meta-data are present.

Other approach would be to introduce a SPI, that would supply metadata (in general) for an image/icon. ImageUtilities could ensure the SPIs would be called just once/instance (as the images should be immutable anyway). Then wrappers could just vanish from UIDefaultsIconMetadata. The drawback would be that if one instance is shared into multiple keys, the metadata (icon id) provider would just pick one of the keys as an identification - I don't know how much could that be an issue.

Reply:
> Other approach would be to introduce a SPI, that would supply metadata (in general) for an image/icon
That sounds a bit backwards. "Metadata" here just means "logical icon ID", right? And surely it's the site that actually loads the icon that should provide the logical icon ID to look up by?

By the way, do we have an actual use case for this patch yet, i.e. actual icons that will be substituted in a future patch? Which icons are those, and how many of them are there?

Reply:
> Oh, I always assumed that this PR was part of the effort to let icons be branded for LAF purposes. Is there any hope of satisfying both needs with one API change?

@eirikbakke I think so, but not necessarily with this PR.  And the question of keeping the metadata in the icon I can understand given the Swing internals at play too.  API discussions are probably better on dev@

> That sounds a bit backwards. "Metadata" here just means "logical icon ID", right? And surely it's the site that actually loads the icon that should provide the logical icon ID to look up by?

I agree with that.  IMO we need an API whereby icon loading can use a logical ID and request an icon for X.  And an SPI that provides the icons for X.  In the LSP _client_ you'd implement the same. Any such API/SPI could use the existing URLs (file locations) as a backup ID to aid transition.  So both needs would benefit from migrating to a logical ID, as well as being able to query that ID, without any icon being loaded.

Reply:
> IMO we need an API whereby icon loading can use a logical ID and request an icon for X. And an SPI that provides the icons for X. In the LSP client you'd implement the same. Any such API/SPI could use the existing URLs (file locations) as a backup ID to aid transition. So both needs would benefit from migrating to a logical ID, as well as being able to query that ID, without any icon being loaded.

Yes, something like this would be exactly what I'd have in mind for this.

> that would require to change UIManager.getIcon|Image to "something different" that would involve that loading SPI (an implementation could do UIManager.get* and return the image/icon with 'logical ID' ) - the change would need to be done throughout the codebase, right ?

I think the existing APIs could be just fine initially. Just keep loading icons by URL; after all, there needs to be a default icon file in the NetBeans codebase for each logical icon ID in any case. But there could indeed be an SPI that ImageUtilities will check to see if a given URL should be resolved to a different Icon instance.

> Good; and what should we do with Swing calling UIManager.get() from its own code ?

For this, just modify the icon in UIDefaults, like Swing intended. We already do this e.g. from Windows8LFCustoms.

Reply:
@sdedic I somewhat suspect there will be fallbacks very much along the lines of what you do here too! I only mentioned in the first place so we sync up conversations on the various uses for IDs / don't lock its spec to something that can't meet more needs in future.